### PR TITLE
#7844: Add ability to gather noc xfer data through dprint

### PR DESCRIPTION
**Pull request by tt-dma**
Passing CI: https://github.com/tenstorrent/tt-metal/actions/runs/9071514938

Ran into an issue where dprint being killed on mmio chip caused hangs (I'm guessing due to prints stacking up on mmio chip when it's no longer being drained), so add a fix for that here as well

### CODE REVIEW

File: tt_metal/hw/inc/dataflow_api.h [hunk 1]
```diff
@@ -582,8 +581,11 @@ void noc_async_read_one_packet_with_state(std::uint32_t src_noc_addr, std::uint3
 
     DEBUG_STATUS("NARW");
 
-    // TODO: need a way sanitize size + addr w/o directly providing x/y here (grab x/y form state?)
-    // DEBUG_SANITIZE_READ_NOC_TRANSACTION(src_noc_addr, dst_local_l1_addr, size);
+    // In order to sanitize, need to grab full noc addr + xfer size from state.
+    std::uint64_t src_noc_addr_full = NOC_CMD_BUF_READ_REG(noc_index, NCRISC_RD_CMD_BUF, NOC_TARG_ADDR_MID);
```
Comment (pgkeller) [inline tt_metal/hw/inc/dataflow_api.h:585]: These need to be wrapped in an ifdef - they are volatile register reads which the compiler can't optimize away

**Review by pgkeller — CHANGES_REQUESTED** [review]

File: tt_metal/hw/inc/dataflow_api.h (hunk 1 above)
Comment (tt-dma) [reply tt_metal/hw/inc/dataflow_api.h:587]: Added an ifdef here, new CI running: https://github.com/tenstorrent/tt-metal/actions/runs/9084172911

**Comment by tt-dma** [comment]
New CI passing: https://github.com/tenstorrent/tt-metal/actions/runs/9084172911

File: tt_metal/hw/inc/dataflow_api.h (hunk 1 above)
Comment (pgkeller) [reply tt_metal/hw/inc/dataflow_api.h:587]: sorry, I should have been more clear, I'm going to nitpick: to keep the mainline code clean/simple, let's create a DEBUG_SANITIZE_NOC_READ_XYZSTATESOMETHING() that handles this and goes away w/ watcher disabled.  Generally I like to push ifdefs down into inline fns or macros so it is easy to see the flow at the top level

**Review by pgkeller — APPROVED** [review]

File: tt_metal/hw/inc/dataflow_api.h (hunk 1 above)
Comment (tt-dma) [reply tt_metal/hw/inc/dataflow_api.h]: Makes sense, updated this. New CI: https://github.com/tenstorrent/tt-metal/actions/runs/9100552147

**Review by pgkeller — APPROVED** [review]

**Comment by tt-dma** [comment]
CI passeth: https://github.com/tenstorrent/tt-metal/actions/runs/9100552147

**Pull request merged**